From fd437302da479ac619c2e3682c835aee3555212c Mon Sep 17 00:00:00 2001
From: Andrii Cherniavskyi <achernia@qti.qualcomm.com>
Date: Mon, 5 Dec 2022 15:38:36 +0100
Subject: [PATCH 5/5] libcamera: pipeline: vivid: Fix format config validation.

Set up and adjust stride, frameSize and colorSpace stream
configuration parameters in validate() function.

Signed-off-by: Andrii Cherniavskyi <achernia@qti.qualcomm.com>
---
 src/libcamera/pipeline/vivid/vivid.cpp | 29 +++++++++++++++++++++-----
 1 file changed, 24 insertions(+), 5 deletions(-)

diff --git a/src/libcamera/pipeline/vivid/vivid.cpp b/src/libcamera/pipeline/vivid/vivid.cpp
index 5728b07a..a59471c2 100644
--- a/src/libcamera/pipeline/vivid/vivid.cpp
+++ b/src/libcamera/pipeline/vivid/vivid.cpp
@@ -72,9 +72,12 @@ public:
 class VividCameraConfiguration : public CameraConfiguration
 {
 public:
-	VividCameraConfiguration();
+	VividCameraConfiguration(VividCameraData *data);
 
 	Status validate() override;
+
+private:
+	VividCameraData *data_;
 };
 
 class PipelineHandlerVivid : public PipelineHandler
@@ -105,8 +108,8 @@ private:
 	}
 };
 
-VividCameraConfiguration::VividCameraConfiguration()
-	: CameraConfiguration()
+VividCameraConfiguration::VividCameraConfiguration(VividCameraData *data)
+	: CameraConfiguration(), data_(data)
 {
 }
 
@@ -135,6 +138,22 @@ CameraConfiguration::Status VividCameraConfiguration::validate()
 
 	cfg.bufferCount = 4;
 
+	V4L2DeviceFormat format;
+	format.fourcc = data_->video_->toV4L2PixelFormat(cfg.pixelFormat);
+	format.size = cfg.size;
+
+	int ret = data_->video_->tryFormat(&format);
+	if (ret)
+		return Invalid;
+
+	cfg.stride = format.planes[0].bpl;
+	cfg.frameSize = format.planes[0].size;
+
+	if (cfg.colorSpace != format.colorSpace) {
+		cfg.colorSpace = format.colorSpace;
+		status = Adjusted;
+	}
+
 	return status;
 }
 
@@ -146,9 +165,9 @@ PipelineHandlerVivid::PipelineHandlerVivid(CameraManager *manager)
 std::unique_ptr<CameraConfiguration>
 PipelineHandlerVivid::generateConfiguration(Camera *camera, Span<const StreamRole> roles)
 {
-	std::unique_ptr<CameraConfiguration> config =
-	       std::make_unique<VividCameraConfiguration>();
 	VividCameraData *data = cameraData(camera);
+	std::unique_ptr<CameraConfiguration> config =
+	       std::make_unique<VividCameraConfiguration>(data);
 
 	if (roles.empty())
 		return config;
-- 
2.49.0

